### DISCUSSION TITLE
How to build OpenBLAS for baremetal ARM Cortex-A9 (ARMv7) on Windows?

### DISCUSSION THREAD

**Discussion by AvihooI** [OpenMathLib/OpenBLAS#4845] [General]
I've been using MSYS2 with the gcc arm none toolchain to compile OpenBLAS for ARMv7 using: `make TARGET=ARMV7 CC=arm-none-eabi-gcc HOSTCC=gcc USE_THREAD=0 EMBEDDED=1 NO_FORTRAN=1 NO_SHARED=1`.

The compilation seems to go fine but it fails at linkage because it is missing system call definitions (read, write, exit, and what have you). 

Why does it even needs those? And regardless, shouldn't these functions be defined by the application that links the library? Is there anything I'm missing here?

**Comment by martin-frbg** [comment]
I have no experience with building for "bare metal" hosts, but I suspect this may be a problem with your toolchain not supplying suitable replacements - when building with EMBEDDED=1, the only external components required should be related to memory management (malloc/free). Another explanation could be that you did a different build in the same path before and did not run `make clean`. There are a small number of code paths that require reading configuration values, or exiting after writing an error message, but offhand I'd think these all involve either multithreading, multi-target (DYNAMIC_ARCH)  builds or cpu affinity mapping

**Reply by AvihooI** [reply]
Okay apparently I needed to give it CROSS=1 for cross compilation. Which I guess makes sense, because otherwise the compiler expects system calls by default.

**Reply by martin-frbg** [reply]
Whoops that should normally be automatic - guess you are actually using an ARM host for building, so the c_check script cannot figure out that you are cross-compiling ? I'll look into making "EMBEDDED=1" another reason why the build should be "CROSS"

**Reply by AvihooI** [reply]
No actually I am using normal Windows x86_64 with MSYS2... however this isn't the first time I noticed this quirky behavior in build systems.

Basically, the linker (LD) needs to be told explicity that there's cross compilation or it demands system call definitions. I don't know what to make of it since I'm no build expert... I just experiment with it until I get the right configuration in place.
